tools/qfvis/source/flow.c

```diff
@@ -300,6 +300,234 @@ free_winding_memory (threaddata_t *thread, size_t winding_mark)
+
+		target_winding = target_portal->winding;
+		target_winding = ClipWinding (thread, target_winding, source_plane, false);
+		if (!target_winding)
+			continue;
+
+		if (!pass_winding) {
+			// unreachable
+		}
+
+		target_winding = ClipWinding (thread, target_winding, pass_plane, false);
+		if (!target_winding)
+			continue;
+
+		// copy source_winding because it likely is already a copy and thus
+		// if it gets clipped away, earlier stack levels will get corrupted
+		source_winding = CopyWinding (thread, prevstack->source_winding);
+
+		source_winding = ClipWinding (thread, source_winding, backplane, false);
+		if (!source_winding) {
+			continue;
+		}
+
+		thread->stats.portaltest++;
+		thread->stats.targettested++;
+
+		if (level > 0) {
+			winding_t  *old = target_winding;
+			if (!separators[0])
+				separators[0] = FindSeparators (thread,
+												source_winding,
+												source_plane,
+												pass_winding, 0);
+			target_winding = ClipToSeparators (thread,
+											   separators[0],
+											   target_winding);
+			if (!target_winding) {
+				thread->stats.targetclipped++;
+				continue;
+			}
+			if (target_winding != old)
+				thread->stats.targettrimmed++;
+		}
+
+		if (level > 1) {
+			winding_t  *old = target_winding;
+			if (!separators[1])
+				separators[1] = FindSeparators (thread,
+												pass_winding,
+												pass_plane,
+												source_winding, 1);
+			target_winding = ClipToSeparators (thread,
+											   separators[1],
+											   target_winding);
+			if (!target_winding) {
+				thread->stats.targetclipped++;
+				continue;
+			}
+			if (target_winding != old)
+				thread->stats.targettrimmed++;
+		}
+
+		thread->stats.sourcetested++;
+		if (level > 2) {
+			winding_t  *old = source_winding;
+			sep_t      *sep;
+			sep = FindSeparators (thread,
+								  target_winding, target_portal->plane,
+								  pass_winding, 0);
+			source_winding = ClipToSeparators (thread, sep, source_winding);
+			free_separators (thread, sep);
+			if (!source_winding) {
+				thread->stats.sourceclipped++;
+				continue;
+			}
+			if (source_winding != old)
+				thread->stats.sourcetrimmed++;
+		}
+
+		if (level > 3) {
+			winding_t  *old = source_winding;
+			sep_t      *sep;
+			sep = FindSeparators (thread, pass_winding, pass_plane,
+								  target_winding, 1);
+			source_winding = ClipToSeparators (thread, sep, source_winding);
+			free_separators (thread, sep);
+			if (!source_winding) {
+				thread->stats.sourceclipped++;
+				continue;
+			}
+			if (source_winding != old)
+				thread->stats.sourcetrimmed++;
+		}
+
+		thread->stats.portalpass++;
+
+		set_add(mightsee, target_portal->cluster);
+		set_add(portalbits, i);
+	}
+
+	free_winding_memory (thread, winding_mark);
+	if (prevstack->next == NULL) {
+		// only free if they aren't borrowed
+		free_separators (thread, separators[1]);
+		free_separators (thread, separators[0]);
+	}
+
+	// replace with reduced version
```

Comment:
Does this rotate the sets? Seems to be a double-buffer that I assume is to avoid the expense of copying the (potentially large) set (hint: better comment). My first thought on this was it's a leak, but then I noticed how `scratchbits` is used (and why `mightsee` can change between iterations).